File: package/ios/Types/VideoStabilizationMode.swift
//
//  VideoStabilizationMode.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 13.10.23.
//  Copyright © 2023 mrousavy. All rights reserved.
//

import AVFoundation
import Foundation

enum VideoStabilizationMode: String, JSUnionValue {
  case off
  case standard
  case cinematic
  case cinematicExtended = "cinematic-extended"
  case auto

  init(jsValue: String) throws {
    if let parsed = VideoStabilizationMode(rawValue: jsValue) {
      self = parsed
    } else {
      throw CameraError.parameter(.invalid(unionName: "videoStabilizationMode", receivedValue: jsValue))
    }
  }

  init(from mode: AVCaptureVideoStabilizationMode) {
    switch mode {
    case .off:
      self = .off
    case .standard:
      self = .standard
    case .cinematic:
      self = .cinematic
    case .cinematicExtended:
      self = .cinematicExtended
    case .auto:
      self = .auto
    default:
      self = .off
    }
  }

  var jsValue: String {
    return rawValue
  }
}

File: package/android/src/main/java/com/mrousavy/camera/core/types/VideoStabilizationMode.kt
package com.mrousavy.camera.core.types

import com.mrousavy.camera.core.InvalidTypeScriptUnionError

enum class VideoStabilizationMode(override val unionValue: String) : JSUnionValue {
  OFF("off"),
  STANDARD("standard"),
  CINEMATIC("cinematic"),
  CINEMATIC_EXTENDED("cinematic-extended");

  private val score: Int
    get() {
      return when (this) {
        OFF -> 0
        STANDARD -> 1
        CINEMATIC -> 2
        CINEMATIC_EXTENDED -> 3
      }
    }

  fun isAtLeast(mode: VideoStabilizationMode): Boolean = score >= mode.score

  companion object : JSUnionValue.Companion<VideoStabilizationMode> {
    override fun fromUnionValue(unionValue: String?): VideoStabilizationMode =
      when (unionValue) {
        "off" -> OFF
        "auto" -> OFF
        "standard" -> STANDARD
        "cinematic" -> CINEMATIC
        "cinematic-extended" -> CINEMATIC_EXTENDED
        else -> throw InvalidTypeScriptUnionError("videoStabilizationMode", unionValue)
      }
  }
}
